src/net.cpp

```diff
@@ -912,6 +912,420 @@ size_t V1Transport::GetSendMemoryUsage() const noexcept
     return m_message_to_send.GetMemoryUsage();
 }
 
+V2Transport::V2Transport(NodeId nodeid, bool initiating, int type_in, int version_in) noexcept :
+    m_cipher{}, m_initiating{initiating}, m_nodeid{nodeid},
+    m_recv_type{type_in}, m_recv_version{version_in},
+    m_recv_state{RecvState::KEY},
+    m_send_state{SendState::KEY}
+{
+    // Initialize the send buffer with ellswift pubkey.
+    m_send_buffer.resize(EllSwiftPubKey::size());
+    std::copy(std::begin(m_cipher.GetOurPubKey()), std::end(m_cipher.GetOurPubKey()), MakeWritableByteSpan(m_send_buffer).begin());
+}
+
+V2Transport::V2Transport(NodeId nodeid, bool initiating, int type_in, int version_in, const CKey& key, Span<const std::byte> ent32) noexcept :
+    m_cipher{key, ent32}, m_initiating{initiating}, m_nodeid{nodeid},
+    m_recv_type{type_in}, m_recv_version{version_in},
+    m_recv_state{RecvState::KEY},
+    m_send_state{SendState::KEY}
+{
+    // Initialize the send buffer with ellswift pubkey.
+    m_send_buffer.resize(EllSwiftPubKey::size());
+    std::copy(std::begin(m_cipher.GetOurPubKey()), std::end(m_cipher.GetOurPubKey()), MakeWritableByteSpan(m_send_buffer).begin());
+}
+
+void V2Transport::SetReceiveState(RecvState recv_state) noexcept
+{
+    AssertLockHeld(m_recv_mutex);
+    // No-op if no change is desired.
+    if (recv_state == m_recv_state) return;
+    // Enforce allowed state transitions.
+    switch (recv_state) {
+    case RecvState::KEY:
+        // Initial state, cannot be transitioned into.
+        Assume(false);
+        break;
+    case RecvState::GARB_GARBTERM:
+        Assume(m_recv_state == RecvState::KEY);
+        break;
+    case RecvState::GARBAUTH:
+        Assume(m_recv_state == RecvState::GARB_GARBTERM);
+        break;
+    case RecvState::VERSION:
+        Assume(m_recv_state == RecvState::GARBAUTH);
+        break;
+    case RecvState::APP:
+        Assume(m_recv_state == RecvState::VERSION || m_recv_state == RecvState::APP_READY);
+        break;
+    case RecvState::APP_READY:
+        Assume(m_recv_state == RecvState::APP);
+        break;
+    }
+    // Change state.
+    m_recv_state = recv_state;
+}
+
+void V2Transport::SetSendState(SendState send_state) noexcept
+{
+    AssertLockHeld(m_send_mutex);
+    // No-op if no change is desired.
+    if (send_state == m_send_state) return;
+    // Enforce allowed state transitions.
+    switch (send_state) {
```

Comment:
nit: in `Set{Send,Receive}State`, at least for my brain the state transition checks would be slightly easier to grasp if the current state is switch/case-d and the allowed next state(s) are `Assume`d (i.e. asserting "where can I go to?" rather than "where could I have come from?").